src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -372,7 +372,113 @@ protected void InitializeReferences()
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                runner.Scan(span);
+
+                Match? match = runner.runmatch;
+                // if we got a match, set runmatch to null if quick is true
+                if (match!._matchcount[0] > 0)
+                {
+                    runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+                    if (match.Text != input)
+                    {
+                        match.Text = input;
+                    }
+
+                    if (quick)
+                    {
+                        runner.runmatch!.Text = null; // Drop reference to text
+                        return null;
+                    }
+
+                    runner.runmatch = null;
+                    match.Tidy(runner.runtextpos);
+
+                    // If there was a match and the original text was sliced, then add beginning to the index to get the real
+                    // Index of the match.
+                    if (match.Success && beginning != 0)
+                    {
+                        match.AddBeginningToIndex(beginning);
+                    }
+
+                    return match;
+                }
+
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+                if (!quick)
+                {
+                    runner.runmatch = null;
+                }
+                else
+                {
+
+                    if (runner.runmatch != null)
+                    {
+                        runner.runmatch.Text = null;
+                    }
+                }
+
+                return RegularExpressions.Match.Empty;
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        internal Match? Run(ReadOnlySpan<char> input, int beginning, int length, int startat)
+        {
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
+            }
+            if ((uint)length > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length, ExceptionResource.LengthNotNegative);
+            }
+
+            RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
+            try
+            {
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.Slice(beginning, length);
+                runner.InitializeForScan(this, span, startat - beginning, quick: true);
+
+                runner.Scan(span);
+
+                Match? match = runner.runmatch;
+                // if we got a match, set runmatch to null if quick is true
+                if (match!._matchcount[0] > 0)
```

Comment:
I see, I guess I never really explored that option of having Scan return a bool, I can do a quick prototype and see if perf is slightly better. For a change like that, and since this API is expected to only be used by us, I suppose it's fine to change return type without having to go through API Review again?

Reply:
> For a change like that, and since this API is expected to only be used by us, I suppose it's fine to change return type without having to go through API Review again?

Yup.  It also needn't be in this PR.